---
title: "Hands-on 16S based microbial diversity analysis using QIIME2"
author: "Belen Carbonetto"
date: "May 2018"
output: html_document
---

## Hands on for 16S based microbial diversity analysis using QIIME2

### Learning outcomes:

**During this hands-on session you will learn how to:**

1- Do quality filtering, denoising and picking of features and representative sequences using DADA2

2- Assign taxonomy to features with trained classifiers.

3- Align sequences and build a phylogenetic tree.

4- Calculate alpha and beta diversity.

5- Test for differential abundances between groups of samples using ANCOM

**Note:** QIIME2 has already been installed in your working station in order to follow this tutorial.

If you are interested in how to install QIIME2 on your own computer please follow [this](https://docs.qiime2.org/2018.2/install/) link.

### Core workflow:

We are going to follow this analysis workflow:

![workflow](https://github.com/mbcarbonetto/qiime2-session/blob/master/hands_on_qiime2.png)

### 1.Quality filtering, denoising and feature picking using DADA2

